README.md

```diff
@@ -1 +1,122 @@
 # ChunkedBase.jl
+
+The package handles the ingestion of data chunks and the distribution & synchronization of work that happens on these chunks in parallel. It came into existence while refactoring the [`ChunkedCSV.jl`](https://github.com/RelationalAI/ChunkedCSV.jl) and [`ChunkedJSONL.jl`](https://github.com/RelationalAI/ChunkedJSONL.jl) packages and was designed to be extended by packages like these. It is a package used to write parser packages.
+
+Specifically, `ChunkedBase.jl` spawns one task which handles IO and behaves as a coordinator, and a configurable number of worker tasks.
+Both CSV and JSONL are textual formats that delimit records by newlines which makes newlines an ideal point to distribute work. One the coordinator task we ingest bytes into a preallocated buffer and then use [`NewlineLexers.jl`](https://github.com/JuliaData/NewlineLexers.jl) package to quickly find newlines in it. In turn, these newlines are distributed among worker tasks and the coordinator immediately starts working on a secondary buffer, while the first one is being processed by the workers. The coordinator task switches back and forth between these two buffers so that lexing can always proceed while the workers are consuming results, ensuring that there is never a gap in available lexed data to consume, keeping the throughput maximized over all N tasks. We refer to this as double-buffering.
+
+The process looks something like this, with the coordinator task at the bottom:
+
+| ![Diagram](/docs/diagrams/chunked_base.png) |
+|:--:|
+| *The coordinator synchronizes with workers using a counter behind a mutex (there is one per buffer). It splits the newlines into N segments, distributes them, and increments the counter by N. After the coordinator distributes work, it starts to process the second chunk of data, while the first buffer is being worked on. There is a handoff happening between the two buffers -- we need to copy the bytes after the last newline from the first buffer to the second. Each worker decrements after the `consume!` is done, and the coordinator will wait for the counter to reach 0 before it overwrites the buffer with new data.* |
+
+Packages like `ChunkedCSV.jl` and `ChunkedJSONL.jl` hook into this structure by defining their own `populate_result_buffer!` methods that parse the records they were assigned into their custom `Result` buffers which are then handed to the `consume!` method (e.g. to be inserted into a database).
+
+The main entry point of this package is the `parse_file_parallel` function, which accepts several "context" arguments, each controlling a different aspect of the process:
+```julia
+function parse_file_parallel(
+    lexer::NewlineLexers.Lexer,
+    parsing_ctx::AbstractParsingContext,
+    consume_ctx::AbstractConsumeContext,
+    chunking_ctx::ChunkingContext,
+    result_buffers::Vector{<:AbstractResultBuffer},
```

Comment:
the example below is great, because i was most confused by `result_buffers`, and now i think i understand what they are... but i don't understand any of the other things 😆 

i think we maybe need an example for each one of 
- parsing
- consuming
- chunking
- result buffers (this one is done)

Reply:
In this example:
* "parsing" is what happens in `populate_result_buffer!`, here we are just copying a newline segment into the result buffer. In real applications, this is where you'd use e.g. Parsers.jl on the `bytes` (the 4th argument), and your `Options` struct would be available in the `parsing_ctx`
* "consuming" is what happens in `consume!`, here we just print what is in the result buffer. In RAICode, this is where you'd push the parsed values into LeafWriters.
* "chunking" is reading data from the input IO object into a preallocated buffer that lives in one of the chunking contexts.

Does this help? I was hoping the README would convey this general idea...

Reply:
i think it would be good to introduce these terms in the README, and what they mean in the context of this package

Reply:
I've added a bit about this, thanks!